qa/tasks/cephfs/test_quiesce.py

```diff
@@ -0,0 +1,389 @@
+import re
+import errno
+import json
+import logging
+import secrets
+import uuid
+from io import StringIO
+import os.path
+from time import sleep
+
+from teuthology.contextutil import safe_while
+from teuthology.exceptions import CommandFailedError
+
+from tasks.cephfs.cephfs_test_case import CephFSTestCase, classhook
+from tasks.cephfs.filesystem import FileLayout, FSMissing
+from tasks.cephfs.fuse_mount import FuseMount
+from tasks.cephfs.caps_helper import (CapTester, gen_mon_cap_str,
+                                      gen_osd_cap_str, gen_mds_cap_str)
+
+log = logging.getLogger(__name__)
+
+INODE_RE = re.compile(r'\[inode 0x([0-9a-fA-F]+)')
+
+# MDS uses linux defines:
+S_IFMT   = 0o0170000
+S_IFSOCK =  0o140000
+S_IFLNK  =  0o120000
+S_IFREG  =  0o100000
+S_IFBLK  =  0o060000
+S_IFDIR  =  0o040000
+S_IFCHR  =  0o020000
+S_IFIFO  =  0o010000
+S_ISUID  =  0o004000
+S_ISGID  =  0o002000
+S_ISVTX  =  0o001000
+
+class TestQuiesce(CephFSTestCase):
+    """
+    Tests for quiescing subvolumes.
+    """
+
+    CLIENTS_REQUIRED = 2
+    MDSS_REQUIRED = 1
+
+    QUIESCE_SUBVOLUME = "subvol_quiesce"
+
+    def setUp(self):
+        super().setUp()
+        self.run_ceph_cmd(f'fs subvolume create {self.fs.name} {self.QUIESCE_SUBVOLUME} --mode=777')
+        p = self.run_ceph_cmd(f'fs subvolume getpath {self.fs.name} {self.QUIESCE_SUBVOLUME}', stdout=StringIO())
+        self.mntpnt = p.stdout.getvalue().strip()
+        self.subvolume = os.path.normpath(os.path.join(p.stdout.getvalue().strip(), ".."))
+
+    def tearDown(self):
+        # restart fs so quiesce commands clean up and commands are left unkillable
+        self.fs.fail()
+        self.fs.set_joinable()
+        self.fs.wait_for_daemons()
+        super().tearDown()
+
+    def _configure_subvolume(self):
+        for m in self.mounts:
+            m.umount_wait()
+        for m in self.mounts:
+            m.update_attrs(cephfs_mntpt = self.mntpnt)
+            m.mount()
+
+    def _client_background_workload(self):
+       SCRIPT = """
+set -ex
+pushd `mktemp -d -p .`
+cp -a /usr .
+popd
+"""
+       for m in self.mounts:
+           p = m.run_shell_payload(SCRIPT, wait=False, stderr=StringIO())
+           m.background_procs.append(p)
+
+    def _wait_for_quiesce_complete(self, reqid):
+        with safe_while(sleep=1, tries=120, action='wait for quiesce completion') as proceed:
+            while proceed():
+                op = self.fs.get_op(reqid)
+                log.debug(f"op:\n{op}")
+                self.assertEqual(op['type_data']['op_name'], 'quiesce_subvolume')
+                if op['type_data']['flag_point'] == 'quiesce complete':
+                    return op
+
+    # TODO: multi-rank
+    def _verify_quiesce(self):
```

Comment:
I wonder if we could add a higher-level validation to try to execute some op against a quiesced inode, including opening new paths under the quiesced root.
Would you expect this to be a different kind of test, or something we could bundle in with this verify?

Reply:
probably a different test